src/parser/parser.go

```diff
@@ -0,0 +1,1228 @@
+		p.endTokenStack.Push(endTokens)
+		defer p.endTokenStack.Pop()
+	}
+
+	flushData := func() {
+		if len(dataBuffer) > 0 {
+			lineno := dataBuffer[0].GetLineno()
+			body = append(body, &nodes.Output{
+				Nodes:      dataBuffer,
+				NodeCommon: nodes.NodeCommon{Lineno: lineno},
+			})
+			dataBuffer = make([]nodes.Node, 0)
+		}
+	}
+
+	for p.stream.Bool() {
+		token := p.stream.Current()
+		if token.Type == lexer.TokenData {
+			if token.Value != "" {
+				// type assert is safe, because token.Type == lexer.TokenData
+				addData(&nodes.TemplateData{
+					Data:       token.Value.(string),
+					NodeCommon: nodes.NodeCommon{Lineno: token.Lineno},
+				})
+			}
+			p.stream.Next()
+		} else if token.Type == lexer.TokenVariableBegin {
+			p.stream.Next()
+			tuple, err := p.parseTuple(false, true, nil, false)
+			if err != nil {
+				return nil, err
+			}
+			addData(tuple)
+			if _, err := p.stream.Expect(lexer.TokenVariableEnd); err != nil {
+				return nil, err
+			}
+		} else if token.Type == lexer.TokenBlockBegin {
+			flushData()
+			p.stream.Next()
+			if endTokens != nil && p.stream.Current().TestAny(endTokens...) {
+				return body, nil
+			}
+			rvs, err := p.parseStatement()
+			if err != nil {
+				return nil, err
+			}
+			body = append(body, rvs...)
+			if _, err := p.stream.Expect(lexer.TokenBlockEnd); err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, p.fail("internal parsing error", nil)
+		}
+	}
+
+	flushData()
+	return body, nil
+}
+
+func (p *parser) parseTuple(simplified bool, withCondexpr bool, extraEndRules []string, explicitParentheses bool) (nodes.Node, error) {
+	lineno := p.stream.Current().Lineno
+	var parse func() (nodes.Node, error)
+	if simplified {
+		parse = p.parsePrimary
+	} else {
+		parse = func() (nodes.Node, error) {
+			return p.parseExpression(withCondexpr)
+		}
+	}
+
+	var args []nodes.Node
+	var isTuple bool
+
+	for {
+		if len(args) > 0 {
+			if _, err := p.stream.Expect(lexer.TokenComma); err != nil {
+				return nil, err
+			}
+		}
+		if p.isTupleEnd(extraEndRules) {
+			break
+		}
+
+		arg, err := parse()
+		if err != nil {
+			return nil, err
+		}
+		args = append(args, arg)
+		if p.stream.Current().Type == lexer.TokenComma {
+			isTuple = true
+		} else {
+			break
+		}
+		lineno = p.stream.Current().Lineno
+	}
+
+	if !isTuple {
+		if len(args) > 0 {
+			return args[0], nil
+		}
+		// if we don't have explicit parentheses, an empty tuple is
+		// not a valid expression.  This would mean nothing (literally
+		// nothing) in the spot of an expression would be an empty
+		// tuple.
+		if !explicitParentheses {
+			return nil, p.fail(fmt.Sprintf("Expected an expression, got %s", p.stream.Current()), nil)
+		}
+	}
+
+	return &nodes.Tuple{
+		Items:      args,
+		Ctx:        "load",
+		NodeCommon: nodes.NodeCommon{Lineno: lineno},
+	}, nil
+}
+
+func (p *parser) parsePrimary() (nodes.Node, error) {
+	token := p.stream.Current()
+	var node nodes.Node
+	if token.Type == lexer.TokenName {
```

Comment:
swicth is better match here especially that simetime one branch correspondds to two different tokens